### PR TITLE
Allow retries to be configurable

### PR DESCRIPTION
**Pull request by jszwedko**
Merges `master` into https://github.com/DataDog/serilog-sinks-datadog-logs/pull/58 and resolves conflicts. I couldn't push to the contributor branch.

Closes: https://github.com/DataDog/serilog-sinks-datadog-logs/pull/58

### CODE REVIEW

File: src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogHttpClient.cs [hunk 1]
```diff
@@ -22,36 +22,35 @@ public class DatadogHttpClient : IDatadogClient
...
         }
 
-        private List<JsonPayloadBuilder> BuildEvents(IEnumerable<LogEvent> events)
+        private List<JsonPayloadBuilder> BuildEvents(IReadOnlyCollection<LogEvent> events)
         {
             var builders = new List<JsonPayloadBuilder>();
             var builder = new JsonPayloadBuilder();
 
+            var chunkBuffer = new List<string>(events.Count());
+            var logEvents = new List<LogEvent>(events.Count());
```
Comment (StephenWakely) [inline src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogHttpClient.cs:53]:
```suggestion
```

These are not being used anywhere. The change from `IEnumerable` to `IReadOnlyCollection` does not seem necessary.

File: src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogSink.cs [hunk 2]
```diff
@@ -100,12 +100,13 @@ public async Task EmitBatchAsync(IEnumerable<LogEvent> events)
         {
             try
             {
-                if (!events.Any())
+                var batch = events.ToArray();
+                if (!batch.Any())
```
Comment (StephenWakely) [inline src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogSink.cs:104]: Likewise, these changes are not needed.

File: src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogTcpClient.cs [hunk 3]
```diff
@@ -87,7 +82,7 @@ private async Task ConnectAsync()
             }
         }
 
-        public async Task WriteAsync(IEnumerable<LogEvent> events)
+        public async Task WriteAsync(IReadOnlyCollection<LogEvent> events)
```
Comment (StephenWakely) [inline src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogTcpClient.cs:85]: This change is unnecessary.

File: src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/IDatadogClient.cs [hunk 4]
```diff
@@ -16,7 +15,7 @@ public interface IDatadogClient
         /// Send payload to Datadog logs-backend.
         /// </summary>
         /// <param name="events">Serilog events to send.</param>
-        Task WriteAsync(IEnumerable<LogEvent> events);
+        Task WriteAsync(IReadOnlyCollection<LogEvent> events);
```
Comment (StephenWakely) [inline src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/IDatadogClient.cs:18]: Likewise.

File: tests/Serilog.Sinks.Datadog.Logs.Tests/NoopClient.cs [hunk 5]
```diff
@@ -22,7 +22,7 @@ public NoopClient(string apiKey, DatadogLogRenderer formatter)
             _formatter = formatter;
         }
 
-        public Task WriteAsync(IEnumerable<LogEvent> events)
+        public Task WriteAsync(IReadOnlyCollection<LogEvent> events)
```
Comment (StephenWakely) [inline tests/Serilog.Sinks.Datadog.Logs.Tests/NoopClient.cs:25]: Likewise.

**Review by StephenWakely — CHANGES_REQUESTED** [review]
Tested and it works, but it looks like a bunch of unused changes have also crept in.

File: src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogTcpClient.cs (hunk 3 above)
Comment (jszwedko) [reply src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogTcpClient.cs:85]: Yeah, I think you are right, it was extra. It was in the contributor PR, but I'm not sure why. I'll remove the unneeded changes.

**Review by StephenWakely — APPROVED** [review]

**Pull request merged**